lib/Optimizer/GraphOptimizer.cpp

```diff
@@ -205,6 +205,155 @@ static Node *simplifyNode(Node *node, Function *F) {
+
+/// fills \p dimGroups with the groups of dimensions of the result of \p node
+/// that get separated. This grouping is only possible if \p node is the first
+/// Reshape node in a ChannelShuffle operation (Reshape->Tranpose->Reshape)
+/// \returns true if successfully grouped output dimensions and false otherwise
+static bool
+groupDimensionsFromChannelShuffleReshape(const ReshapeNode &node,
+                                         ChannelShuffleDimGroups &dimGroups) {
+  const auto dims = node.getDims();
+  const auto inputDims = node.getInput().dims();
+  if (dims.size() != inputDims.size() + 1) {
+    return false;
+  }
+  dimGroups.resize(inputDims.size());
+  bool foundReshapedDim = false;
+  for (size_t dimIdx = 0, groupIdx = 0; dimIdx < dims.size();
+       ++dimIdx, ++groupIdx) {
+    auto &group = dimGroups[groupIdx];
+    group.push_back(dimIdx);
+    if (std::find(inputDims.begin(), inputDims.end(), dims[dimIdx]) ==
+        inputDims.end()) {
+      // make sure only one input dimension is changed and there's at least one
+      // more dim
+      if (foundReshapedDim || dimIdx >= dims.size() - 1) {
+        return false;
+      }
+      group.push_back(++dimIdx);
+      foundReshapedDim = true;
+    }
+  }
+  return true;
+}
+
+// Sink Transpose below ChannelShuffle node sequence. For example
+// (Transpose_1->Reshape_1->Transpose_2->Reshape_2) becomes
+// (Reshape_1->Transpose_2->Reshape_2->Transpose_1) and Reshape_1,
+// Transpose_2, Reshape_2 are reconstructed so that the end result is the
+// ChannelShuffle operation on the same dimention as before Transpose_1 was
+// sunk.
+static bool sinkTranposeBelowChannelShuffle(Function *F, Node *node) {
+  auto *postShuffleRN = dyn_cast<ReshapeNode>(node);
+  if (!postShuffleRN) {
+    return false;
+  }
+
+  auto *shuffleTR = dyn_cast<TransposeNode>(postShuffleRN->getInput());
+  if (!shuffleTR) {
+    return false;
+  }
+
+  auto *preShuffleRN = dyn_cast<ReshapeNode>(shuffleTR->getInput());
+  if (!preShuffleRN) {
+    return false;
+  }
+
+  auto *sinkingTR = dyn_cast<TransposeNode>(preShuffleRN->getInput());
+  if (!sinkingTR) {
+    return false;
+  }
+
+  // Try to group the dimensions of preShuffleRN by which input dimension they
+  // came from. This should always succeed for the first ReshapeNode in a
+  // ChannelShuffle.
+  ChannelShuffleDimGroups preShuffleRNDimGroups;
+  if (!groupDimensionsFromChannelShuffleReshape(*preShuffleRN,
+                                                preShuffleRNDimGroups)) {
+    return false;
+  }
+
+  // Get the inverse of the shuffle applied by the TranposeNode that is being
+  // sunk in order to apply this to the dimensions of preShuffleRN.
+  auto sinkingTRShuffle = sinkingTR->getShuffle();
+  llvm::SmallVector<unsigned_t, max_tensor_dimensions> inverseSinkingTRShuffle(
+      sinkingTRShuffle.size());
+  for (auto i = 0; i < sinkingTRShuffle.size(); i++) {
```

Comment:
the convention we follow is to call size once, see loops here:
https://llvm.org/docs/CodingStandards.html#turn-predicate-loops-into-predicate-functions